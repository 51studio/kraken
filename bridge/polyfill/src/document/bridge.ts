import { krakenJSToDart } from '../kraken';

// Timeout for batch updater, default to 60 fps.
const FRAME_TICK_TIMEOUT = 1000 / 60;

// Auto negotiation whether to enable batch update.
let isEnableBatchUpdate:boolean = false;
let frameTimeoutTimer:any;

export function enableBatchUpdate() {
  isEnableBatchUpdate = true;
}

const tickMessageQueue:string[] = [];

export function frameTick() {
  if (frameTimeoutTimer !== undefined) clearTimeout(frameTimeoutTimer);
  frameTimeoutTimer = setTimeout(frameTick, FRAME_TICK_TIMEOUT);
  if (tickMessageQueue.length > 0) {
    krakenJSToDart('["batchUpdate",[' + tickMessageQueue.join(',') + ']]');
    tickMessageQueue.length = 0;
  }
}

export function createElement(type: string, id: number, props: any, events: any) {
  const message = `["createElement", [{"id":${id},"type":"${type}","props":${JSON.stringify(props)},"events":${JSON.stringify(events)}}]]`;
  if (isEnableBatchUpdate) {
    tickMessageQueue.push(message);
  } else {
    krakenJSToDart(message);
  }
}

export function createTextNode(id: number, nodeType: number, data: string) {
  const message = `["createTextNode",[{"id":${id},"nodeType":${nodeType},"data":"${data}"}]]`;
  if (isEnableBatchUpdate) {
    tickMessageQueue.push(message);
  } else {
    krakenJSToDart(message);
  }
}

export function insertAdjacentNode(parentNodeId: number, position: string, nodeId: number) {
  const message = `["insertAdjacentNode",[${parentNodeId},"${position}",${nodeId}]]`;
  if (isEnableBatchUpdate) {
    tickMessageQueue.push(message);
  } else {
    krakenJSToDart(message);
  }
}

export function removeNode(id: number) {
  const message = `["removeNode",[${id}]]`;
  if (isEnableBatchUpdate) {
    tickMessageQueue.push(message);
  } else {
    krakenJSToDart(message);
  }
}

export function setProperty(id: number, key: string, value: any) {
  const message = `["setProperty",[${id},"${key}","${value}"]]`;
  if (isEnableBatchUpdate) {
    tickMessageQueue.push(message);
  } else {
    krakenJSToDart(message);
  }
}

export function setStyle(id: number, key: string, value: string) {
  const message = `["setStyle",[${id},"${key}","${value}"]]`;
  if (isEnableBatchUpdate) {
    tickMessageQueue.push(message);
  } else {
    krakenJSToDart(message);
  }
}

export function addEvent(id: number, eventName: string) {
  const message = `["addEvent",[${id},"${eventName}"]]`;
  if (isEnableBatchUpdate) {
    tickMessageQueue.push(message);
  } else {
    krakenJSToDart(message);
  }
}

export function removeEvent(id: number, eventName: string) {
  const message = `["removeEvent",[${id},"${eventName}"]]`;
  if (isEnableBatchUpdate) {
    tickMessageQueue.push(message);
  } else {
    krakenJSToDart(message);
  }
}
