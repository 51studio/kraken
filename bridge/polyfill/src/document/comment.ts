import { NodeImpl, NodeType } from './node';
import { createComment } from './UIManager';

export class CommentImpl extends NodeImpl {
  public nodeName: string = '#comment';
  public data: string = '';

  constructor(data: string) {
    super(NodeType.COMMENT_NODE);
    this.textContent = this.nodeValue = this.data = data;
    createComment(this.nodeId, data);
  }
}
