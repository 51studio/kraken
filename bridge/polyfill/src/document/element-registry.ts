import { AudioElement } from './elements/audio';
import { VideoElement } from './elements/video';
import { CanvasElement } from './elements/canvas';
import { ImageElement } from './elements/img';
import { IframeElement } from './elements/iframe';
import { AnimationPlayerElement } from './elements/animation-player';
import { AnchorElement } from "./elements/anchor";

const ElementRegistryMap = {};

// TODO: export registry by [window.customElements]
export const ElementRegistry = {
  define(name: string, constructor: any) {
    ElementRegistryMap[name] = constructor;
  },
  get(name: string) {
    return ElementRegistryMap[name];
  },
  // TODO: support while needing it.
  // whenDefined(name: string) {},
};

ElementRegistry.define('audio', AudioElement);
ElementRegistry.define('video', VideoElement);
ElementRegistry.define('canvas', CanvasElement);
ElementRegistry.define('iframe', IframeElement);
ElementRegistry.define('img', ImageElement);
ElementRegistry.define('a', AnchorElement);
ElementRegistry.define('animation-player', AnimationPlayerElement);
