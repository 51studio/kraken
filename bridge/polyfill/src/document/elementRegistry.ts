import { AudioElement } from './elements/audio';
import { VideoElement } from './elements/video';
import { CanvasElement } from './elements/canvas';

const ElementRegistryMap = {};

// TODO: export registry by [window.customElements]
export const ElementRegistry = {
  define(name: string, constructor: any) {
    ElementRegistryMap[name] = constructor;
  },
  get(name: string) {
    return ElementRegistryMap[name];
  },
  // TODO: support while needing it.
  // whenDefined(name: string) {},
};

ElementRegistry.define('audio', AudioElement);
ElementRegistry.define('video', VideoElement);
ElementRegistry.define('canvas', CanvasElement);
