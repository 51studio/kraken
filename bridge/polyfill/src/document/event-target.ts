import { addEvent } from './ui-manager';

export const BODY = -1;
// Window is not inherit node but EventTarget, so we assume window is a node.
export const WINDOW = -2;

type EventHandler = EventListener;

export class EventTarget {
  public nodeId: number;
  __eventHandlers: Map<string, Array<EventHandler>> = new Map();
  __propertyEventHandler: Map<string, EventHandler> = new Map();

  constructor(nodeId?: number, builtInEvents: Array<string> = []) {
    if (nodeId) {
      this.nodeId = nodeId;
    }
    builtInEvents.forEach(event => {
      let eventName = 'on' + event.toLowerCase();
      Object.defineProperty(this, eventName, {
        get() {
          return this.__propertyEventHandler.get(event);
        },
        set(fn: EventHandler) {
          const preHandler = this.__propertyEventHandler[event];
          this.removeEventListener(event, preHandler);
          this.__propertyEventHandler.set(event, fn);
          if (typeof fn === 'function') {
            this.addEventListener(event, fn);
          }
        }
      });
    });
  }

  public addEventListener(eventName: string, handler: EventHandler) {
    if (typeof handler !== 'function') {
      return;
    }
    if (!this.__eventHandlers.has(eventName) || this.nodeId === BODY) {
      this.__eventHandlers.set(eventName, []);

      // this is an bargain optimize for addEventListener which send `addEvent` message to kraken Dart side only once and no one can stop element to
      // trigger event from dart side. this can led to significant performance improvement when using Front-End frameworks such as Rax, or cause some
      /// overhead performance issue when some event trigger more frequently.
      if (this.nodeId) {
        addEvent(this.nodeId, eventName);
      }
    }
    this.__eventHandlers.get(eventName)!.push(handler);
  }

  // Do not really emit remove event, due to performance consideration.
  public removeEventListener(eventName: string, handler: EventHandler) {
    if (typeof handler !== 'function' || !this.__eventHandlers.has(eventName)) {
      return;
    }
    let newHandler = this.__eventHandlers.get(eventName)!.filter(fn => fn != handler);
    this.__eventHandlers.set(eventName, newHandler);
  }

  public dispatchEvent(event: Event) {
    if (!this.__eventHandlers.has(event.type)) {
      return;
    }
    event.currentTarget = event.target = this;
    let stack = this.__eventHandlers.get(event.type)!.slice();

    for (let i = 0; i < stack.length; i++) {
      stack[i].call(this, event);
    }

    return !event.defaultPrevented;
  }
}

export class Event {
  type: string;
  cancelable: boolean;
  currentTarget: EventTarget;
  target: EventTarget;
  defaultPrevented: boolean;

  [key: string]: any;

  constructor(type: string) {
    this.type = type;
  }
}
