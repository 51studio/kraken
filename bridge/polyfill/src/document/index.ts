
import { addKrakenUIListener } from '../bridge';
import { krakenUIListener } from './ui-listener';

addKrakenUIListener(krakenUIListener);

export { document } from './document';
