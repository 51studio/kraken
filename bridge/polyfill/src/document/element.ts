import { BODY } from './events/event-target';
import { Node, NodeType, traverseNode } from './node';
import { document } from './document';

import {
  createElement,
  setProperty,
  removeProperty,
  setStyle,
  method,
  toBlob,
  getProperty
} from './ui-manager';

const RECT_PROPERTIES = [
  'offsetTop',
  'offsetLeft',
  'offsetWidth',
  'offsetHeight',

  'clientWidth',
  'clientHeight',
  'clientLeft',
  'clientTop',

  'scrollTop',
  'scrollLeft',
  'scrollHeight',
  'scrollWidth',
];

interface ICamelize {
  (str: string): string;
}

/**
 * Create a cached version of a pure function.
 */
function cached(fn: ICamelize) {
  const cache = Object.create(null);
  return function cachedFn(str : string) {
    const hit = cache[str];
    return hit || (cache[str] = fn(str));
  };
};

/**
 * Camelize a hyphen-delimited string.
 */
const camelize: ICamelize = (str: string) => {
  const camelizeRE = /-(\w)/g;
  return str.replace(camelizeRE, function (_: string, c: string) {
    return c ? c.toUpperCase() : '';
  });
}

// Cached camelize utility
const cachedCamelize = cached(camelize);
// support event handlers using 'on' property prefix.
const elementBuildInEvents = ['click', 'appear', 'disappear', 'touchstart', 'touchmove', 'touchend', 'touchcancel'];

class StyleDeclaration {
  private targetId: number;
  constructor(targetId: number) {
    this.targetId = targetId;
  }
  setProperty(property: string, value: any) {
    const camelizedProperty = cachedCamelize(property);
    this[camelizedProperty] = value;
    setStyle(this.targetId, camelizedProperty, value);
  }
  removeProperty(property: string) {
    const camelizedProperty = cachedCamelize(property);
    setStyle(this.targetId, camelizedProperty, '');
    const originValue = this[camelizedProperty];
    this[camelizedProperty] = '';
    return originValue;
  }
  getPropertyValue(property: string) {
    const camelizedProperty = cachedCamelize(property);
    return this[camelizedProperty];
  }
}

export class Element extends Node {
  public readonly tagName: string;
  public style: StyleDeclaration;
  // TODO use NamedNodeMap: https://developer.mozilla.org/en-US/docs/Web/API/NamedNodeMap
  public attributes: Array<any> = [];

  constructor(tagName: string, _targetId?: number, builtInEvents?: Array<string>, builtInProperties?: Array<string>) {
    super(NodeType.ELEMENT_NODE, _targetId, elementBuildInEvents.concat(builtInEvents || []));
    this.tagName = tagName.toUpperCase();
    const targetId = this.targetId;
    const style = this.style = new StyleDeclaration(targetId);
    // FIXME: Proxy not support in iOS 9.x
    // See: https://caniuse.com/#search=proxy
    this.style = new Proxy(style, {
      set(target: any, key: string, value: any, receiver: any): boolean {
        style.setProperty(key, value);
        return true;
      },
      get(target: any, key: string, receiver) {
        // Proxy to prototype method
        if (key in target) {
          return target[key];
        }
        return style.getPropertyValue(key);
      },
    });

    // Define rect properties
    for (let i = 0; i < RECT_PROPERTIES.length; i++) {
      const prop = RECT_PROPERTIES[i];
      Object.defineProperty(this, prop, {
        configurable: false,
        enumerable: true,
        get() {
          return Number(method(targetId, prop));
        },
      });
    }

    if (Array.isArray(builtInProperties)) {
      builtInProperties.forEach(property => {
        Object.defineProperty(this, property, {
          get() {
            return this.getAttribute(property);
          },
          set(value) {
            this.setAttribute(property, value);
          }
        });
      });
    }

    // Body is created automaticlly.
    if (this.targetId != BODY) {
      createElement(this.tagName, targetId);
    }
  }

  getBoundingClientRect = () => {
    const rectInformation = method(this.targetId, 'getBoundingClientRect');
    if (typeof rectInformation === 'string') {
      return JSON.parse(rectInformation);
    } else {
      return null;
    }
  }

  get nodeName() {
    return this.tagName.toUpperCase();
  }

  public setAttribute(name: string, value: string) {
    // The attribute name is automatically converted to
    // all lower-case when setAttribute() is called on an HTML element in an HTML document
    name = String(name).toLowerCase();
    value = String(value);
    if (this.attributes[name]) {
      this._didModifyAttribute(name, this.attributes[name].value, value);
      this.attributes[name].value = value;
    } else {
      const attr = {name, value};
      this.attributes[name] = attr;
      this.attributes.push(attr);
    }
    setProperty(this.targetId, name, value);
  }

  otifyNodeRemoved(insertionNode: Node) {
    if (insertionNode.isConnected) {
      traverseNode(this, (node:Node) => {
        if (node instanceof Element) {
          node.notifyChildRemoved();
        }
      });
    }
  }
  notifyChildRemoved() {
    const elementid = this.getAttribute('id');
    if (elementid !== null && elementid !== undefined && elementid !== '') {
      this._updateId(elementid, '');
    }
  }

  notifyNodeInsert(insertionNode: Node) :void {
    if (insertionNode.isConnected) {
      const elementid = this.getAttribute('id');
      if (elementid !== null && elementid !== undefined && elementid !== '') {
        this._updateId('', elementid);
      }
    }
  }


  private _didModifyAttribute(name:string, oldValue:string, newValue:string) :void {
    if (name === 'id') {
      this._checkupdateId(oldValue, newValue);
    }
  }

  private _checkupdateId(oldValue:string, newValue:string) :void {
    if (!this.isConnected) {
      return;
    }
    if (oldValue === newValue) {
      return;
    }
    if (oldValue !== '' && oldValue !== undefined && oldValue !== null) {
      this._updateId(oldValue, newValue);
    }
  }
  private _updateId(oldValue:string, newValue:string) {
    if (oldValue !== null && oldValue !== undefined && oldValue !== '') {
      document.removeElementById(oldValue, this);
    }
    if (newValue !== null && newValue !== undefined && newValue !== '') {
      document.addElementById(newValue, this);
    }
  }

  public getAttribute(name: string) {
    name = String(name);
    if (this.attributes[name]) {
      return this.attributes[name].value;
    }
    return getProperty(this.targetId, name);
  }

  public hasAttribute(name: string) {
    name = String(name);
    return Boolean(this.attributes[name]);
  }

  public removeAttribute(name: string) {
    if (this.attributes[name]) {
      const attr = this.attributes[name];
      const idx = this.attributes.indexOf(attr);
      if (idx !== -1) {
        this.attributes.splice(idx, 1);
      }
      this._didModifyAttribute(name, attr.value, '');

      removeProperty(this.targetId, name);
      delete this.attributes[name];
    }
  }

  public click() {
    method(this.targetId, 'click');
  }

  async toBlob(devicePixelRatio: number = window.devicePixelRatio) {
    return toBlob(this.targetId, devicePixelRatio);
  }

  public scroll(x: number | any, y?: number) {
    let option = x;
    if (typeof x === 'number') {
      option = {
        'top': y,
        'left': x
      };
    }
    method(this.targetId, 'scroll', [option]);
  }

  public scrollTo(x: number | any, y?: number) {
    if (typeof y === 'number') {
      scroll(x, y);
    } else {
      scroll(x, 0);
    }
  }

  public scrollBy(x: number | any, y?: number) {
    let option = x;
    if (typeof x === 'number') {
      option = {
        'top': y,
        'left': x
      };
    }
    method(this.targetId, 'scrollBy', [option]);
  }
}
