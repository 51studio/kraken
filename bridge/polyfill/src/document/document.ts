import { Node, NodeType, NodeId, traverseNode } from './node';
import { Element } from './element';
import { Comment } from './comment';
import { TextNode } from './text';
import { ElementRegistry } from './element-registry';

export class Document extends Node {
  private bodyElement = new Element('BODY', NodeId.BODY);
  public body: Element = this.bodyElement;
  // @TODO Need to implement complete document tree model, equal to body temporary
  public documentElement: Element = this.bodyElement;
  public nodeName: string = '#document';
  public nodeType = NodeType.DOCUMENT_NODE;

  constructor() {
    // Use the same nodeId with body, only used in event targets,
    // document events are triggered and received by body element.
    super(NodeType.DOCUMENT_NODE, NodeId.BODY);
  }

  createElement(tagName: string) : Element {
    const ElementConstructor = ElementRegistry.get(tagName) || Element;
    return new ElementConstructor(tagName);
  }

  createTextNode(text: string) {
    return new TextNode(text);
  }

  /**
   * createComment() creates a new comment node, and returns it.
   * @param data {string} A string containing the data to be added to the Comment.
   */
  createComment(data: string) {
    return new Comment(data);
  }
}

export const document = new Document();

export function getNodeByNodeId(nodeId: number) : Node|null {
  let _node = null;
  traverseNode(document.body, (node: Node) : any => {
    if (node.nodeId === nodeId) {
      _node = node;
      return true; // Return true to stop traversing
    }
  });
  return _node;
}
