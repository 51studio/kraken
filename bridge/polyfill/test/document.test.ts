import '../src/document/index';

describe('document', () => {
  let messages: string[] = [];
  beforeEach(() => {
    messages = [];
    // @ts-ignore
    global.__kraken_js_to_dart__ = (msg) => messages.push(msg);
  });

  it('createElement', () => {
    var div = document.createElement('div');
    var span = document.createElement('span');
    var textNode = document.createTextNode('helloworld');
    span.appendChild(textNode);
    div.appendChild(span);
    document.body.appendChild(div);
    expect(messages).toMatchSnapshot('createElement');
    expect(div.parentNode!.nodeName).toMatch('BODY');
    expect(span.parentNode!.nodeName).toMatch('DIV');
    expect(textNode.parentNode!.nodeName).toMatch('SPAN');
  });

  it('insertBefore', () => {
    var div = document.createElement('div');
    var span = document.createElement('span');
    var textNode = document.createTextNode('helloworld');
    span.appendChild(textNode);
    div.appendChild(span);
    document.body.appendChild(div);

    var insertText = document.createTextNode('inserted');
    var insertSpan = document.createElement('span');
    insertSpan.appendChild(insertText);
    div.insertBefore(insertSpan, span);
    expect(messages).toMatchSnapshot('insertBefore');
  });

  it('setStyle', () => {
    let div = document.createElement('div');
    div.style.fontSize = '12px';
    expect(messages).toMatchSnapshot('setStyle');
  });
});
