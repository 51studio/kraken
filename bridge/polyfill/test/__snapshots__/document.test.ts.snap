// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`document createElement: createElement 1`] = `
Array [
  "[\\"createElement\\", [{
    \\"type\\": \\"DIV\\",
    \\"id\\": 1,
    \\"props\\": {},
    \\"events\\": []
  }]]",
  "[\\"createElement\\", [{
    \\"type\\": \\"SPAN\\",
    \\"id\\": 2,
    \\"props\\": {},
    \\"events\\": []
  }]]",
  "[\\"createTextNode\\", [{
    \\"id\\": 3,
    \\"nodeType\\": 3,
    \\"data\\": \\"helloworld\\"
  }]]",
  "[\\"insertAdjacentNode\\", [2, \\"beforeend\\", 3]]",
  "[\\"insertAdjacentNode\\", [1, \\"beforeend\\", 2]]",
  "[\\"insertAdjacentNode\\", [-1, \\"beforeend\\", 1]]",
]
`;

exports[`document insertBefore: insertBefore 1`] = `
Array [
  "[\\"createElement\\",[{\\"type\\":\\"DIV\\",\\"id\\":1,\\"props\\":{},\\"events\\":[]}]]",
  "[\\"createElement\\",[{\\"type\\":\\"SPAN\\",\\"id\\":2,\\"props\\":{},\\"events\\":[]}]]",
  "[\\"createTextNode\\",[{\\"id\\":3,\\"nodeType\\":3,\\"data\\":\\"helloworld\\"}]]",
  "[\\"insertAdjacentNode\\",[2,\\"beforeend\\",3]]",
  "[\\"insertAdjacentNode\\",[1,\\"beforeend\\",2]]",
  "[\\"insertAdjacentNode\\",[-1,\\"beforeend\\",1]]",
  "[\\"createTextNode\\",[{\\"id\\":4,\\"nodeType\\":3,\\"data\\":\\"inserted\\"}]]",
  "[\\"createElement\\",[{\\"type\\":\\"SPAN\\",\\"id\\":5,\\"props\\":{},\\"events\\":[]}]]",
  "[\\"insertAdjacentNode\\",[5,\\"beforeend\\",4]]",
  "[\\"insertAdjacentNode\\",[2,\\"beforebegin\\",5]]",
]
`;

exports[`document setStyle: setStyle 1`] = `
Array [
  "[\\"createElement\\",[{\\"type\\":\\"DIV\\",\\"id\\":1,\\"props\\":{},\\"events\\":[]}]]",
  "[\\"setProperty\\",[1,\\"style\\",{\\"fontSize\\":\\"12px\\"}]]",
]
`;
